File: backend-workers/src/routes/leagues.ts
import { Hono } from 'hono';
import { Env, LeagueSchema, TeamSchema } from '../models/schemas';
import { DatabaseQueries } from '../db/queries';
import { YahooFantasyService } from '../services/yahoo';
import { CacheService } from '../services/cache';
import { refreshAccessToken } from './oauth';

const leagues = new Hono<{ Bindings: Env }>();

leagues.post('/', async (c) => {
  try {
    const leagueId = c.req.query('league_id');
    const season = parseInt(c.req.query('season') || '2024');
    const sessionId = c.req.query('session');

    if (!leagueId) {
      return c.json({ error: 'league_id is required' }, 400);
    }

    const db = new DatabaseQueries(c.env.DB);
    const cache = new CacheService(c.env.CACHE);

    // Get user's OAuth token if session provided
    let accessToken: string | undefined;
    if (sessionId) {
      const sessionData = await c.env.CACHE.get(`session:${sessionId}`, 'json') as any;
      if (sessionData?.user_id) {
        let tokenData = await db.getOAuthToken(sessionData.user_id);

        if (tokenData) {
          // Check if token is expired
          const now = Math.floor(Date.now() / 1000);
          if (tokenData.expires_at < now) {
            // Token expired, refresh it
            const refreshed = await refreshAccessToken(
              tokenData.refresh_token,
              c.env.YAHOO_CLIENT_ID,
              c.env.YAHOO_CLIENT_SECRET
            );

            // Update token in database
            await db.updateOAuthToken(
              sessionData.user_id,
              refreshed.access_token,
              now + refreshed.expires_in
            );

            accessToken = refreshed.access_token;
          } else {
            accessToken = tokenData.access_token;
          }
        }
      }
    }

    const yahooService = new YahooFantasyService(accessToken);

    // Fetch league from Yahoo
    const league = await yahooService.getLeague(leagueId, season);
    if (!league) {
      return c.json({ error: 'League not found or not accessible' }, 404);
    }

    // Validate with Zod
    const validatedLeague = LeagueSchema.parse(league);

    // Store in database
    await db.insertLeague(validatedLeague);

    // Store teams
    for (const team of validatedLeague.teams) {
      await db.insertTeam(team);
    }

    // Cache the league
    await cache.set(cache.generateLeagueKey(leagueId), validatedLeague, 3600);

    return c.json(validatedLeague, 201);
  } catch (error: any) {
    console.error('Error adding league:', error);
    return c.json({ error: error.message || 'Internal server error' }, 500);
  }
});

leagues.get('/', async (c) => {
  try {
    const season = c.req.query('season') ? parseInt(c.req.query('season')!) : undefined;

    const db = new DatabaseQueries(c.env.DB);
    const leagueRows = await db.getAllLeagues(season);

    const leagues = leagueRows.map((row: any) => ({
      league_id: row.league_id,
      name: row.name,
      platform: row.platform,
      season: row.season,
      current_week: row.current_week,
      num_teams: row.num_teams,
      scoring_type: row.scoring_type,
      teams: [],
    }));

    return c.json(leagues);
  } catch (error: any) {
    console.error('Error getting leagues:', error);
    return c.json({ error: error.message || 'Internal server error' }, 500);
  }
});

leagues.get('/:leagueId', async (c) => {
  try {
    const leagueId = c.req.param('leagueId');

    const db = new DatabaseQueries(c.env.DB);
    const cache = new CacheService(c.env.CACHE);

    // Try cache first
    const cacheKey = cache.generateLeagueKey(leagueId);
    const cached = await cache.get(cacheKey);
    if (cached) {
      return c.json(cached);
    }

    // Get from database
    const leagueRow = await db.getLeague(leagueId);
    if (!leagueRow) {
      return c.json({ error: 'League not found' }, 404);
    }

    const teamRows = await db.getTeamsByLeague(leagueId);
    const teams = teamRows.map((t: any) => TeamSchema.parse(t));

    const league = {
      ...leagueRow,
      teams,
    };

    // Cache for 1 hour
    await cache.set(cacheKey, league, 3600);

    return c.json(league);
  } catch (error: any) {
    console.error('Error getting league:', error);
    return c.json({ error: error.message || 'Internal server error' }, 500);
  }
});

leagues.get('/:leagueId/teams', async (c) => {
  try {
    const leagueId = c.req.param('leagueId');

    const db = new DatabaseQueries(c.env.DB);
    const teamRows = await db.getTeamsByLeague(leagueId);
    const teams = teamRows.map((t: any) => TeamSchema.parse(t));

    return c.json(teams);
  } catch (error: any) {
    console.error('Error getting teams:', error);
    return c.json({ error: error.message || 'Internal server error' }, 500);
  }
});

leagues.delete('/:leagueId', async (c) => {
  try {
    const leagueId = c.req.param('leagueId');

    const db = new DatabaseQueries(c.env.DB);
    const cache = new CacheService(c.env.CACHE);

    await db.deleteLeague(leagueId);
    await cache.delete(cache.generateLeagueKey(leagueId));

    return c.json({ message: 'League deleted successfully' });
  } catch (error: any) {
    console.error('Error deleting league:', error);
    return c.json({ error: error.message || 'Internal server error' }, 500);
  }
});

export default leagues;

File: backend/app/api/leagues.py
from fastapi import APIRouter, HTTPException, Query
from typing import List, Optional
from ..models.league import League, Team
from ..services.yahoo import YahooFantasyService
from ..db import db

router = APIRouter()


@router.post("/", response_model=League)
async def add_league(
    league_id: str = Query(..., description="Yahoo league ID"),
    season: int = Query(2024, description="Season year")
):
    """
    Add a new league to track
    Fetches league data from Yahoo and stores it
    """
    yahoo_service = YahooFantasyService()

    try:
        # Fetch league data from Yahoo
        league = await yahoo_service.get_league(league_id, season)

        if not league:
            raise HTTPException(status_code=404, detail="League not found or not accessible")

        # Store league in database
        await db.execute(
            """
            INSERT OR REPLACE INTO leagues
            (league_id, name, platform, season, current_week, num_teams, scoring_type)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            (
                league.league_id,
                league.name,
                league.platform,
                league.season,
                league.current_week,
                league.num_teams,
                league.scoring_type
            )
        )

        # Store teams
        for team in league.teams:
            await db.execute(
                """
                INSERT OR REPLACE INTO teams
                (team_id, league_id, name, manager_name, wins, losses, ties, points_for, points_against)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    team.team_id,
                    team.league_id,
                    team.name,
                    team.manager_name,
                    team.wins,
                    team.losses,
                    team.ties,
                    team.points_for,
                    team.points_against
                )
            )

        return league

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    finally:
        await yahoo_service.close()


@router.get("/", response_model=List[League])
async def get_leagues(season: Optional[int] = None):
    """Get all tracked leagues"""
    try:
        if season:
            rows = await db.fetch_all(
                "SELECT * FROM leagues WHERE season = ? ORDER BY created_at DESC",
                (season,)
            )
        else:
            rows = await db.fetch_all(
                "SELECT * FROM leagues ORDER BY created_at DESC"
            )

        leagues = []
        for row in rows:
            league = League(
                league_id=row["league_id"],
                name=row["name"],
                platform=row["platform"],
                season=row["season"],
                current_week=row["current_week"],
                num_teams=row["num_teams"],
                scoring_type=row["scoring_type"]
            )
            leagues.append(league)

        return leagues

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{league_id}", response_model=League)
async def get_league(league_id: str):
    """Get a specific league"""
    try:
        row = await db.fetch_one(
            "SELECT * FROM leagues WHERE league_id = ?",
            (league_id,)
        )

        if not row:
            raise HTTPException(status_code=404, detail="League not found")

        # Get teams
        team_rows = await db.fetch_all(
            "SELECT * FROM teams WHERE league_id = ?",
            (league_id,)
        )

        teams = [
            Team(
                team_id=t["team_id"],
                league_id=t["league_id"],
                name=t["name"],
                manager_name=t["manager_name"],
                wins=t["wins"],
                losses=t["losses"],
                ties=t["ties"],
                points_for=t["points_for"],
                points_against=t["points_against"]
            )
            for t in team_rows
        ]

        league = League(
            league_id=row["league_id"],
            name=row["name"],
            platform=row["platform"],
            season=row["season"],
            current_week=row["current_week"],
            num_teams=row["num_teams"],
            scoring_type=row["scoring_type"],
            teams=teams
        )

        return league

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{league_id}/teams", response_model=List[Team])
async def get_league_teams(league_id: str):
    """Get all teams in a league"""
    try:
        rows = await db.fetch_all(
            "SELECT * FROM teams WHERE league_id = ? ORDER BY wins DESC, points_for DESC",
            (league_id,)
        )

        teams = [
            Team(
                team_id=row["team_id"],
                league_id=row["league_id"],
                name=row["name"],
                manager_name=row["manager_name"],
                wins=row["wins"],
                losses=row["losses"],
                ties=row["ties"],
                points_for=row["points_for"],
                points_against=row["points_against"]
            )
            for row in rows
        ]

        return teams

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{league_id}")
async def delete_league(league_id: str):
    """Remove a league from tracking"""
    try:
        await db.execute(
            "DELETE FROM leagues WHERE league_id = ?",
            (league_id,)
        )
        return {"message": "League deleted successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
